### ISSUE TITLE
Don't bundle next install it within the app generator

### ISSUE THREAD

**Issue by ryardley** [blitz-js/blitz#471]
### What do you want and why?

A common usecase for blitz is to enhance an existing Next app and add blitz manually. Out of this I found a couple of things that are preventing this workflow. 

What I want to do:

1. Take an existing next app
2. `yarn add blitz --dev`
3. Have the blitz commands all work.

I noticed a couple of things break if I tried to do this this is one of them

---

Currently we are bundling next as a dependency of the @blitzjs/server package which might be a mistake as it limits the use of blitz when enhancing an existing next app

### Possible implementation(s)

Move next to a peer dependency and add next to the app generator

### Additional context

This would be a breaking change as it will mean existing apps would need to install next separately.

**Comment by flybayer** [comment]
Initially, I really think we should keep bundling Next, but let's discuss it.

Can you explain more in-depth what you mean by "have blitz commands work" in a Next.js app? 

Also, why are you trying to enhance a Next app instead of converting it wholesale to Blitz?

**Comment by aem** [comment]
Bundling Next could have some unintended side effects for this case, so I'm with @ryardley here. If Blitz has a different version of Next pinned than the host app (speaking strictly in the additive case Rudi suggested here) then we could inadvertently either try to use APIs that didn't exist in the host version (not that bad, just weird) or upgrade them to a version where an API they were using has been deprecated + removed (very bad, will be a breaking change to install Blitz). Either way, the user isn't actually using the version of `next` that they have in their `package.json` which would be confusing.

I think the best bet is a peer dependency + a preinit hook in the `blitz` command that checks the installed version of Next to make sure it satisfies our peer dep. Normally I'd say `npm`/`yarn`'s warnings are enough, but since we will absolutely break if the peer dep isn't met I think an extra safeguard is probably a good idea

**Comment by flybayer** [comment]
If we want to support Blitz as an addon to Next, then yes I agree it needs to be peer dependency.

But I'm wondering why would we support Blitz as an addon to Next?

**Comment by aem** [comment]
Isn't that effectively the upgrade path to converting an existing `next` app to a `blitz` app?

**Comment by flybayer** [comment]
No. As of today, the upgrade path is:

1. Remove Next dependency
2. Add Blitz dependency
3. Move `/pages/api/` to `/api/`
4. Rename `next.config.js` to `blitz.config.js`

**Comment by wKovacs64** [comment]
Current upgrade path sounds almost like an installer. `blitz install blitz` 😉

**Comment by ryardley** [comment]
> Can you explain more in-depth what you mean by "have blitz commands work" in a Next.js app?

We should be able to turn a next app into a blitz app with as few changes to the codebase as possible. In my head that could be by simply running `npm install blitz` and then changing scripts within `package json`. 

Once that is done then `blitz start` and `blitz build` should work as expected alongside all other cli commands. `next.config.js` can be converted automatically to `blitz.config.js` on `blitz build` or `blitz start`. The developer can then slowly start colocating contexts and structuring their existing apps or creating new models and resolvers. It should also be possible to push the recently converted blitz app to production with no additional deployment changes. 

> Also, why are you trying to enhance a Next app instead of converting it wholesale to Blitz?

tldr; We need to provide as minimal code foot print as possible to drive adoption.

Common usecases for adopting blitz will be 
* "I have a JAM stack style next marketing site but I want to flesh out the backend to turn it into a platform"
* "I have a simple next app that authenticates with firebase but I want to get off it because I am worried about vendor lock as I want full control over the backend"

So the next -> blitz path is really core in terms of our adoption path. We want people to try using Blitz on their existing next apps.  

One of the key reasons why Next.js has been so successful is because of the way that Zeit (now Vercel) respect minimalism in their API and a key example of that is how few changes you need to make in order to manually bootstrap a next app. This makes the mental model of a next app incredibly simple and that is what drives adoption as opposed to simply providing an install script. If an installer adds a bunch of stuff to your codebase that is not really required it  obfuscates the mental model and becomes harder for a developer to understand exactly how the framework works which in-turn makes the framework hard to learn. 

With Blitz we should aim for a super simple zero installer enhancement path because for the same reasons that people choose Next they will choose Blitz.

**Comment by ryardley** [comment]
> I think the best bet is a peer dependency + a preinit hook in the blitz command that checks the installed version of Next to make sure it satisfies our peer dep. Normally I'd say npm/yarn's warnings are enough, but since we will absolutely break if the peer dep isn't met I think an extra safeguard is probably a good idea

We definitely need a runtime check

**Comment by ryardley** [comment]
https://asciinema.org/a/93ebYAbfdfpolhDr9Ey9H7WNa

**Comment by flybayer** [comment]
@ryardley ok yes, I agree 100% with you on it being very easy to convert a next app to a blitz app!

The only thing I'm not convinced of is whether to stop bundling Next with Blitz. 

I think this point has essentially zero affect on the upgrade path, because it's just one more small change that we can make on initial start. I think this is more a technical decision for us than an adoption UX issue.

We can continue bundling Next and still have the workflow you wanted in the OP:

> What I want to do:
> 
> 1. Take an existing next app
> 2. yarn add blitz --dev
> 3. Have the blitz commands all work.

So let's compare:

## Blitz bundles Next

### Pros

- Simpler for blitz users. They only think about Blitz instead of Blitz + Next. And only one version to upgrade
- We never have to worry about a version compatibility bug
- We have absolute control over Next, it's feature flags, it's versions, etc.
- At some point we can maintain a fork of Next if needed for some advanced Blitz features


### Cons

- Users have to wait for a blitz upgrade to use a fancy new feature of Next (but it might not work anyways)
- You have to remove the Next dependency when upgrading from Next to Blitz (I think this is a non-issue)

## Blitz does not bundle Next

### Pros

- Users can upgrade to a Next release before Blitz does. But it's not guaranteed they can use the feature because it may require a change inside Blitz — so another opportunity for bugs and confusion
- Don't have to remove Next as a dependency when upgrading Next to Blitz

### Cons

- Many opportunities for bugs because of the tight Blitz <> Next coupling.
- We have to run tests against multiple Next versions.
- High likelyhood of users being confused: see that Next is a dependency, so go read the Next documentation but then the Next docs are going to mislead them.

Assuming a I haven't missed anything major (please let me know if so), I think there's overwhelming reason to keep bundling Next. We can always re-evaluate this down the road, so there's no need to change until we hit something that is like, yes, we obviously need to make this change.

What do you think?

**Comment by ryardley** [comment]
I agree supporting multiple next versions could be problematic. Honestly my issue is that it should be a one step install process which has as few code changes as possible. I would be happy if we pin the peerDependency to a single version and running `blitz start` or `blitz build` against an incorrect version simply detected and installed the correct version and wouldn’t run without it. That way we win in a few ways. We get a small initial download as well as a simple and clear installation process that can be done manually without any codemods, only need to support a single version and can open up the range at a later point if we need or see fit?

**Comment by flybayer** [comment]
Good point, the download size is definitely a pro/con! I added that to my pro/con list.

Enforcing a single Next version is a good idea. But this is still a bad UX for users. Because at first glance it _appears_ you can change the next version, but then you get an unpleasant surprise when you run `blitz start`. As much as possible, we need to make impossible things impossible.

I can see that having Next as a peer dependency is a small benefit for migrating to Blitz, but then isn't it a step backwards for the majority of Blitz users who will start from scratch with Blitz?

In other words, I think we should prioritize the UX of greenfield Blitz apps, when there is a conflict between greenfield and migration UX.

**Comment by flybayer** [comment]
Circling back to this. I think bundling next has worked very well. Do you have any updated thoughts @ryardley?
